scylla/src/transport/load_balancing/latency_aware.rs

```diff
@@ -0,0 +1,253 @@
+        if let Some((last_min_avg, last_update)) = *min_latency {
+            if last_update.elapsed() < self.update_rate {
+                return Some(last_min_avg); // not yet time for updating
+            }
+        }
+
+        let now = Instant::now();
+        let mut min_avg: Option<Duration> = None;
+
+        for node in nodes.iter() {
+            if let Some(avg) = *node.average_latency.read().unwrap() {
+                match *min_latency {
+                    None => {
+                        min_avg = Some(avg.average);
+                        *min_latency = Some((avg.average, now));
+                    }
+                    Some((old_avg, _)) if avg.average < old_avg => {
+                        min_avg = Some(avg.average);
+                        *min_latency = Some((avg.average, now));
+                    }
+                    _ => (),
+                };
+            }
+        }
+
+        min_avg
+    }
+
+    fn make_plan<'a>(&self, plan: Vec<Arc<Node>>) -> Plan<'a> {
+        let min_avg_latency = match self.update_min_avg_latency(&plan) {
+            Some(min_avg) => min_avg,
+            None => return Box::new(plan.into_iter()), // noop, as no latency data has been collected yet
+        };
+
+        Box::new(IteratorWithSkippedNodes::new(
+            Box::new(plan.into_iter()),
+            self.exclusion_threshold,
+            self.retry_period,
+            self.minimum_measurements,
+            min_avg_latency,
+        ))
+    }
+}
+
+impl LoadBalancingPolicy for LatencyAwarePolicy {
+    fn plan<'a>(&self, _statement: &Statement, cluster: &'a ClusterData) -> Plan<'a> {
+        self.make_plan(cluster.all_nodes.clone())
+    }
+
+    fn name(&self) -> String {
+        "LatencyAwarePolicy".to_string()
+    }
+
+    fn requires_latency_measurements(&self) -> bool {
+        true
+    }
+}
+
+impl ChildLoadBalancingPolicy for LatencyAwarePolicy {
+    fn apply_child_policy(
+        &self,
+        plan: Vec<Arc<Node>>,
+    ) -> Box<dyn Iterator<Item = Arc<Node>> + Send + Sync> {
+        self.make_plan(plan)
+    }
+}
+
+enum FastEnough {
+    Yes,
+    No { average: Duration },
+}
+
+enum IteratorWithSkippedNodes<'a> {
+    FirstRun {
+        all_nodes: Plan<'a>,
+        skipped: Vec<Arc<Node>>,
+        minimum_measurements: usize,
+        retry_period: Duration,
+        exclusion_threshold: f64,
+        min_avg: Duration,
+    },
+    SkippedRun {
+        skipped: Plan<'a>,
+    },
+}
+
+impl<'a> IteratorWithSkippedNodes<'a> {
+    fn new(
+        nodes: Plan<'a>,
+        exclusion_threshold: f64,
+        retry_period: Duration,
+        minimum_measurements: usize,
+        min_avg: Duration,
+    ) -> Self {
+        Self::FirstRun {
+            all_nodes: nodes,
+            exclusion_threshold,
+            retry_period,
+            minimum_measurements,
+            min_avg,
+            skipped: vec![],
+        }
+    }
+
+    fn fast_enough(
+        node: &Arc<Node>,
```

Comment:
Nit: you are not cloning the `&Arc<Node>`, so you can as well change the argument type to `&Node` so that this function is able to accept a wider range of arguments. Besides, `&Arc<_>` is a pointer-to-a-pointer, so it may be that the program needs to do double indirection instead of one in order to access the contents (unless the optimizer does its job).